unison-cli/src/Unison/Codebase/Editor/HandleInput/Update.hs

```diff
@@ -189,12 +201,324 @@ handleUpdate input optionalPatch requestedNames = do
         & Path.resolve @_ @_ @Path.Absolute currentPath'
         & tShow
 
+getSlurpResultForUpdate :: Set Name -> Cli r (SlurpResult Symbol)
+getSlurpResultForUpdate requestedNames = do
+  (slurp, termRefToNames, nameToTermRefs) <- do
+    slurpCheckNames <- Branch.toNames <$> Cli.getCurrentBranch0
+
+    let slurp :: TypecheckedUnisonFile Symbol Ann -> SlurpResult Symbol
+        slurp file =
+          Slurp.slurpFile file (Set.map Name.toVar requestedNames) Slurp.UpdateOp slurpCheckNames
+
+    let termRefToNames :: TermReferenceId -> Set Symbol
+        termRefToNames =
+          Set.map Name.toVar . Names.namesForReferent slurpCheckNames . Referent.fromTermReferenceId
+
+    let nameToTermRefs :: Symbol -> Set TermReference
+        nameToTermRefs = Names.refTermsNamed slurpCheckNames . Name.unsafeFromVar
+
+    pure (slurp, termRefToNames, nameToTermRefs)
+
+  slurp1 <- do
+    Cli.Env {codebase} <- ask
+    unisonFile0 <- Cli.expectLatestTypecheckedFile
+
+    -- Here, we identify whether there are any "implicit terms", which are terms that should be brought into the latest
+    -- typechecked Unsion file and re-typechecked, which also re-discovers components.
+    --
+    -- The running example here will be one in which the current namespace already has the following terms stored.
+    -- (For simplicity, we will ignore the limitation that mutually recursive definitions must be top-level arrows;
+    -- pretend this is Haskell's let).
+    --
+    --   ping = pong + 1   (reference = #pingpong.ping)
+    --   pong = ping + 2   (reference = #pingpong.pong)
+    --   wham = pong + 3   (reference = #wham)
+    --
+    -- The user re-defines "ping" in their scratch file thus:
+    --
+    --   ping = wham + 4   (reference = #newping)
+    --
+    -- Note that, pre-update, we had two components [ping,pong] and [wham]. But after the update, since the new `ping`
+    -- refers to the old `wham`, which refers to the old `pong`, which refers to the old `ping`, we really want to end
+    -- up with a single component in the end, [ping,pong,wham].
+
+    -- First, compute an initial slurp, which will identify the initial set of definitions we are updating ({"ping"}).
+    -- Also, this will be the slurp that we fall back on, in case re-typechecking another Unison file with implicit
```

Comment:
Out of curiosity, when can this fail? I haven't made it to the transcripts yet, but hopefully there's a transcript that shows that behaviour 😄

Reply:
This fails when the user makes a type-incompatible update; the existing transcripts do cover that case